pkg/blockbuilder/blockbuilder.go

```diff
@@ -867,54 +848,49 @@ func (b *BlockBuilder) consumePartition(
 			Offset: firstRec.Offset,
 		}
 		b.seekPartition(client, part, rec)
-		return pl, nil
+		return nil
 	}
 
 	// All samples in all records were processed. We can commit the last record's offset.
 	if commitRec == nil {
 		commitRec = lastRec
 	}
 
+	commitRecOffset := commitRec.Offset + 1 // offset+1 means everything up to (including) commitRec was processed
+
 	// If there were records that we consumed but didn't process, we must rewind the partition's offset
 	// to the commit record. This is so on the next cycle, when the partition is read again, the consumer
 	// starts at the commit point.
 	defer func() {
 		rec := kgo.EpochOffset{
 			Epoch:  commitRec.LeaderEpoch,
-			Offset: commitRec.Offset + 1, // offset+1 means everything up (including) to commitRec was processed
+			Offset: commitRecOffset,
 		}
 		b.seekPartition(client, part, rec)
 	}()
 
+	// The new lag is the distance between fully processed offsets.
+	pl.Lag -= commitRecOffset - firstRec.Offset
+
 	// We should take the max of "seen till" offset. If the partition was lagging
 	// due to some record not being processed because of a future sample, we might be
 	// coming back to the same consume cycle again.
-	commitSeenTillOffset := seenTillOffset
-	if lastRec != nil && commitSeenTillOffset < lastRec.Offset {
-		commitSeenTillOffset = lastRec.Offset
+	if pl.SeenTillOffset < lastRec.Offset {
+		pl.SeenTillOffset = lastRec.Offset
 	}
 	// Take the max of block max times because of same reasons above.
-	commitBlockEnd := blockEnd
-	if lastBlockEnd > blockEnd {
-		commitBlockEnd = lastBlockEnd
-	}
-
-	pl = partitionInfo{
-		Partition: pl.Partition,
-		Lag:       lag,
-		Commit: kadm.Offset{
-			Topic:       commitRec.Topic,
-			Partition:   commitRec.Partition,
-			At:          commitRec.Offset + 1,
-			LeaderEpoch: commitRec.LeaderEpoch,
-			Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), commitSeenTillOffset, commitBlockEnd),
-		},
-		CommitRecTs:    commitRec.Timestamp.UnixMilli(),
-		SeenTillOffset: commitSeenTillOffset,
-		LastBlockEnd:   commitBlockEnd,
+	pl.LastBlockEnd = max(blockEnd, lastBlockEnd)
+
+	pl.Commit = kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRecOffset,
+		LeaderEpoch: commitRec.LeaderEpoch,
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), pl.SeenTillOffset, pl.LastBlockEnd),
 	}
+	pl.CommitRecTs = commitRec.Timestamp.UnixMilli()
 
-	return pl, commitOffset(ctx, b.logger, client, b.cfg.Kafka.ConsumerGroup, pl.Commit)
+	return commitOffset(ctx, b.logger, client, b.cfg.Kafka.ConsumerGroup, pl.Commit)
```

Comment:
We should ideally update `pl` only when `commitOffset()` does not return an error. How about we call `commitOffset()` before making changes to `pl`?